File: apps/web/app/api/src/services/ragService.ts
// RAG Service - Retrieves and formats data from JSON file for LLM context
import { StudyTask, ScheduleEvent } from '../types';
import { getUserTasks, getUserSchedule } from './dataService';

/**
 * Format database data into structured text for LLM system prompt
 */
export async function formatContextForLLM(
  tasks: StudyTask[],
  events: ScheduleEvent[],
  currentTime: Date,
  moduleWorkloads?: Record<string, number>
): Promise<string> {
  const contextLines: string[] = [];

  // Format current time
  contextLines.push(`Current Time: ${currentTime.toISOString().replace('T', ' ').slice(0, 19)} UTC`);
  contextLines.push('');

  // Format tasks section
  contextLines.push("=== USER'S STUDY TASKS (from Canvas) ===");
  if (tasks.length > 0) {
    for (const task of tasks) {
      // Try to match course name to module code for workload lookup
      // Simple exact match: course name === module code
      const workload = moduleWorkloads?.[task.course];
      const workloadStr = workload !== undefined ? ` | Workload: ${workload} hours/week` : '';

      contextLines.push(
        `- Task: ${task.title} | Course: ${task.course} | Due: ${task.dueAt}${workloadStr}`
      );
    }
  } else {
    contextLines.push('- No tasks found.');
  }
  contextLines.push('');

  // Format schedule section
  contextLines.push("=== USER'S CLASS SCHEDULE (from NUSMods) ===");
  if (events.length > 0) {
    // Filter out past events on the current day
    const currentDay = currentTime.toLocaleDateString('en-US', { weekday: 'long' });
    const currentHour = currentTime.getHours();
    const currentMinute = currentTime.getMinutes();
    const currentTimeInt = currentHour * 60 + currentMinute;
    
    const upcomingEvents = events.filter(event => {
      const dayOrder = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];
      const currentDayIdx = dayOrder.indexOf(currentDay) >= 0 ? dayOrder.indexOf(currentDay) : 0;
      const eventDayIdx = dayOrder.indexOf(event.day) >= 0 ? dayOrder.indexOf(event.day) : 99;
      
      // Parse event start time (HHMM format)
      const eventStartInt = event.startTime.length === 4
        ? parseInt(event.startTime.slice(0, 2), 10) * 60 + parseInt(event.startTime.slice(2), 10)
        : 0;
      
      const daysUntilEvent = (eventDayIdx - currentDayIdx + 7) % 7;
      
      // Keep future days or today's events that haven't started yet
      return daysUntilEvent > 0 || (daysUntilEvent === 0 && eventStartInt > currentTimeInt);
    });

    if (upcomingEvents.length === 0) {
      contextLines.push('- No upcoming classes found.');
    } else {
      // Group by day
      const eventsByDay: Record<string, ScheduleEvent[]> = {};
      for (const event of upcomingEvents) {
        if (!eventsByDay[event.day]) {
          eventsByDay[event.day] = [];
        }
        eventsByDay[event.day].push(event);
      }

      // Sort days
      const dayOrder = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];
      const sortedDays = Object.keys(eventsByDay).sort(
        (a, b) => (dayOrder.indexOf(a) >= 0 ? dayOrder.indexOf(a) : 99) -
          (dayOrder.indexOf(b) >= 0 ? dayOrder.indexOf(b) : 99)
      );

      for (const day of sortedDays) {
        const dayEvents = eventsByDay[day];
        // Sort events by start time within each day (chronologically)
        dayEvents.sort((a, b) => a.startTime.localeCompare(b.startTime));
        contextLines.push(`\n${day}:`);
        for (const event of dayEvents) {
          const startTime = event.startTime.length === 4
            ? `${event.startTime.slice(0, 2)}:${event.startTime.slice(2)}`
            : event.startTime;
          const endTime = event.endTime.length === 4
            ? `${event.endTime.slice(0, 2)}:${event.endTime.slice(2)}`
            : event.endTime;
          contextLines.push(
            `  - ${event.module} (${event.type}) | ${startTime}-${endTime} | Venue: ${event.venue || 'N/A'}`
          );
        }
      }
    }
  } else {
    contextLines.push('- No schedule events found.');
  }

  return contextLines.join('\n');
}

/**
 * Get user tasks and format context for LLM
 */
export async function getRAGContext(userId: string): Promise<string> {
  const tasks = await getUserTasks(userId);
  const events = await getUserSchedule(userId);
  const currentTime = new Date();

  return formatContextForLLM(tasks, events, currentTime);
}

File: apps/api/src/services/rag_service.py
"""RAG Service - Retrieves and formats data from database for LLM context."""
from typing import List, Optional
from datetime import datetime
from sqlalchemy.orm import Session
from ..models.db_models import StudyTask, ScheduleEvent


def get_user_tasks(db: Session, user_id: str, limit: int = 20) -> List[StudyTask]:
    """Get all tasks for a user, ordered by due date."""
    tasks = db.query(StudyTask).filter(
        StudyTask.user_id == user_id
    ).order_by(StudyTask.due_at.asc()).limit(limit).all()
    return tasks


def get_user_schedule(db: Session, user_id: str, day: Optional[str] = None) -> List[ScheduleEvent]:
    """Get schedule events for a user, optionally filtered by day."""
    query = db.query(ScheduleEvent).filter(ScheduleEvent.user_id == user_id)
    
    if day:
        query = query.filter(ScheduleEvent.day == day)
    
    events = query.order_by(ScheduleEvent.day, ScheduleEvent.start_time).all()
    return events


def format_context_for_llm(
    tasks: List[StudyTask],
    events: List[ScheduleEvent],
    current_time: datetime
) -> str:
    """Format database data into structured text for LLM system prompt."""
    context_lines = []
    
    # Format current time
    context_lines.append(f"Current Time: {current_time.strftime('%Y-%m-%d %H:%M:%S')} UTC")
    context_lines.append("")
    
    # Format tasks section
    context_lines.append("=== USER'S STUDY TASKS (from Canvas) ===")
    if tasks:
        for task in tasks:
            due_date_str = task.due_at
            context_lines.append(
                f"- Task: {task.title} | Course: {task.course} | Due: {due_date_str} | Link: {task.link or 'N/A'}"
            )
    else:
        context_lines.append("- No tasks found.")
    context_lines.append("")
    
    # Format schedule section
    context_lines.append("=== USER'S CLASS SCHEDULE (from NUSMods) ===")
    if events:
        # Group by day
        events_by_day = {}
        for event in events:
            if event.day not in events_by_day:
                events_by_day[event.day] = []
            events_by_day[event.day].append(event)
        
        # Sort days
        day_order = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
        sorted_days = sorted(events_by_day.keys(), key=lambda d: day_order.index(d) if d in day_order else 99)
        
        for day in sorted_days:
            day_events = events_by_day[day]
            context_lines.append(f"\n{day}:")
            for event in day_events:
                start_time = f"{event.start_time[:2]}:{event.start_time[2:]}" if len(event.start_time) == 4 else event.start_time
                end_time = f"{event.end_time[:2]}:{event.end_time[2:]}" if len(event.end_time) == 4 else event.end_time
                context_lines.append(
                    f"  - {event.module} ({event.type}) | {start_time}-{end_time} | Venue: {event.venue or 'N/A'}"
                )
    else:
        context_lines.append("- No schedule events found.")
    
    return "\n".join(context_lines)
